### PR TITLE
pkg/trace/stats/quantile: accept ddsketch distributions

### PR DESCRIPTION
**Pull request by piochelepiotr**
Update the stats endpoint of the agent to accept ddsketches.

We receive 2 ddsketches: one for errors, one for hits. They are serialized with protobuf inside the stats message.

### CODE REVIEW

File: pkg/trace/pb/stats.proto [hunk 1]
```diff
@@ -32,4 +30,7 @@ message ClientGroupedStats {
 	uint64 hits = 7;
 	uint64 errors = 8;
 	uint64 duration = 9;
+	double top_level = 10;
```
Comment (furmmon) [inline pkg/trace/pb/stats.proto:33]: top_level is not set in the tracer

File: pkg/trace/pb/ddsketch.proto [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/* Unless explicitly stated otherwise all files in this repository are licensed under the Apache License 2.0.
```
Comment (richardstartin) [inline pkg/trace/pb/ddsketch.proto:1]: Can't the DDSketch library do all of this for you? I would expect to be able to call `proto.Unmarshal(bytes, &sketch)` and then call [this](https://github.com/DataDog/sketches-go/blob/master/ddsketch/ddsketch.go#L145) in DDSketch

Comment (piochelepiotr) [reply pkg/trace/pb/ddsketch.proto]: Good point about using the library protobuf.
I updated the code to use their struct. But I need to test the performance on that, because it's doing more allocations, and also, this looks really inefficient:
https://github.com/DataDog/sketches-go/blob/master/ddsketch/store/dense_store.go#L166

Comment (richardstartin) [reply pkg/trace/pb/ddsketch.proto]: Indeed. Perhaps we could improve it afterwards though?

File: pkg/trace/agent/agent.go [hunk 3]
```diff
@@ -261,6 +268,92 @@ func (a *Agent) Process(p *api.Payload, sublayerCalculator *stats.SublayerCalcul
...
+
+			tags := map[string]string{"version": in.Version}
+			if b.HTTPStatusCode != 0 {
+				tags["http.status_code"] = strconv.Itoa(int(b.HTTPStatusCode))
+			}
+			newb := stats.Bucket{
+				Start:    int64(group.Start),
+				Duration: int64(group.Duration),
+				Counts:   make(map[string]stats.Count),
+				Distributions: make(map[string]stats.Distribution),
```
Comment (furmmon) [inline pkg/trace/agent/agent.go:298]: ErrDistribution map creation is missing

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	for index := minOffset; index < maxIndex; index++ {
+		gErr := errDDSketch.get(index)
+		gHits := okDDSketch.get(index) + gErr
+		if gHits == 0 {
+			// if gHits == 0, gErr == 0 also
+			continue
+		}
+		hits.N += gHits
+		errors.N += gErr
```
Comment (furmmon) [inline pkg/trace/stats/quantile/ddsketch_converter.go:60]: nit: I expect while reading to find errors.N under gErr check (line 67)

File: pkg/trace/pb/stats.proto [hunk 5]
```diff
@@ -11,14 +9,14 @@ message ClientStatsPayload {
 	string hostname = 1;
 	string env = 2;
 	string version = 3;
-	repeated ClientStatsBucket stats = 4 [(gogoproto.nullable) = false];
+	repeated ClientStatsBucket stats = 4;
```
Comment (furmmon) [inline pkg/trace/pb/stats.proto:12]: why did you change this?
You can generate with gogoproto
// protoc -I=. -I=$GOPATH/src -I=$GOPATH/src/github.com/gogo/protobuf/protobuf --gogofaster_out=. ddsketch.proto stats.proto

**Review by richardstartin — APPROVED** [review]
Approving for compatibility with the tracer

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+			V:     v,
+			G:     gErr,
+			Delta: int(2 * EPSILON * float64(errors.N-1)),
+		})
+	}
+	if hits.N > 0 {
+		hits.Entries[0].Delta = 0
+		hits.Entries[len(hits.Entries)-1].Delta = 0
+	}
+	if errors.N > 0 {
```
Comment (furmmon) [inline pkg/trace/stats/quantile/ddsketch_converter.go:136]: On the corner case of N constant values, I wonder what will happen in further merges.
If N >= 2, there should be at least a min and a max value on the GK implem

**Review by raphaelgavache — APPROVED** [review]
lgtm

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+		count += int(c)
+	}
+	return count
+}
+
+func (s *ddSketch) maxSize() int {
+	return len(s.bins) + len(s.contiguousBins)
+}
+
+func getIndexes(s1 ddSketch, s2 ddSketch) []int {
```
Comment (piochelepiotr) [inline pkg/trace/stats/quantile/ddsketch_converter.go:36]: there is a lot of logic here to make sure we don't put the same index twice in the array.

Another way to do it would be to add a condition `if index == previousIndex { continue }` in the main loop. It might be easier. What do you think?

File: pkg/trace/agent/agent.go [hunk 8]
```diff
@@ -7,6 +7,7 @@ package agent
 
 import (
 	"context"
+	"github.com/DataDog/datadog-agent/pkg/trace/stats/quantile"
```
Comment (gbbr) [inline pkg/trace/agent/agent.go:10]: Make sure to put this in the right group.

File: pkg/trace/agent/agent.go [hunk 9]
```diff
@@ -318,6 +321,26 @@ func (a *Agent) ProcessStats(in pb.ClientStatsPayload, lang string) {
 				TopLevel: float64(b.Hits),
 				Value:    float64(b.Duration),
 			}
+			if hitsSummary, errorSummary, err := quantile.DDToGKSketches(b.HitsSummary, b.ErrorSummary); err == nil {
```
Comment (gbbr) [inline pkg/trace/agent/agent.go:324]: Nit: can we reverse the `if` branches? It's usually `err != nil` first. Also, the variable names seem unnecessarily long, would you mind shortening?
```suggestion
			if hits, errors, err := quantile.DDToGKSketches(b.HitsSummary, b.ErrorSummary); err == nil {
```

File: pkg/trace/pb/README.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
+## Generating Go files from protobuf files
```
Comment (gbbr) [inline pkg/trace/pb/README.md:1]: I'm on the fence about this. Some of the generated code might've been modified. I'm not sure if we want to encourage folk to re-generate this.

File: pkg/trace/pb/ddsketch.proto (hunk 2 above)
Comment (gbbr) [reply pkg/trace/pb/ddsketch.proto:1]: If we can do it using the DDSketch library as suggested, let's drop this protobuf definition. Performance is not _that_ critical for a request coming in once every 10 seconds here, and I agree with Richard that we can improve on the upstream later on if we must. It's best to share this code for obvious reasons.

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
+package quantile
```
Comment (gbbr) [inline pkg/trace/stats/quantile/ddsketch_converter.go:1]: `converter.go` should be enough for the file name. Don't forget to add the copyright headers.

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+import (
+	"errors"
```
Comment (gbbr) [inline pkg/trace/stats/quantile/ddsketch_converter.go:4]: Can you please sort the imports into groups like the rest of the source? We use 3 groups:

1. Stdlib
2. Local (datadog-agent) imports
3. External

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/sketches-go/ddsketch/mapping"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/gogo/protobuf/proto"
+	"sort"
+)
+
+type ddSketch struct {
```
Comment (gbbr) [inline pkg/trace/stats/quantile/ddsketch_converter.go:13]: `gofmt` and some minimal docs would be welcome. Not sure if needed here or if the reader who is aware of DDSketch should automatically understand this. If it's the latter no worries.

File: pkg/trace/stats/quantile/ddsketch_converter.go [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+	case pb.IndexMapping_CUBIC:
+		return mapping.NewCubicallyInterpolatedMappingWithGamma(mappingPb.Gamma, mappingPb.IndexOffset)
+	default:
+		return nil, fmt.Errorf("interpolation not supported: %d", mappingPb.Interpolation)
+	}
+}
+
+// DDToGKSketches converts two dd sketches: ok and errors to 2 gk sketches: hits and errors
+// with hits = ok + errors
+func DDToGKSketches(okSketchData []byte, errSketchData []byte) (hits, errors *SliceSummary, err error) {
```
Comment (gbbr) [inline pkg/trace/stats/quantile/ddsketch_converter.go:118]: Can we have a single func called `DDSketchToGK` which takes only one argument `[]byte` and converts it? It makes more sense imho from an API perspective.

I'm not very familiar with the code, so if they are needed together, ignore me :)

File: pkg/trace/stats/quantile/ddsketch_converter_test.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	for _, v := range append(hits, errors...) {
+		assert.InEpsilon(42.0, v, relativeValueError)
+	}
+}
+
+func TestDDSketchToGKConstant10(t *testing.T) {
+	testDDSketchToGKConstant(t, 10)
+}
+
+func TestDDSketchToGKConstant1e3(t *testing.T) {
```
Comment (gbbr) [inline pkg/trace/stats/quantile/ddsketch_converter_test.go:81]: Let's have just one test with multiple sub-tests.

**Review by gbbr — COMMENTED** [review]
Left a few nits, hopefully helping you move closer to a mergeable state. Please request another review once you clean up debugging code and move it out of draft and I'll do another...

File: pkg/trace/stats/quantile/ddsketch_converter.go (hunk 14 above)
Comment (piochelepiotr) [reply pkg/trace/stats/quantile/ddsketch_converter.go:118]: That would be ideal yes. But hits = ok + errors, so we need to sum them at the same time as converting them, hence the function that take the two.
An alternative would be to:
1. Decode to ddsketch
2. Sum ok and errors
3. Convert hits and errors to gksketch

But that leads to more allocations (to sum ok and errors)

File: pkg/trace/pb/ddsketch.proto (hunk 2 above)
Comment (piochelepiotr) [reply pkg/trace/pb/ddsketch.proto:1]: That makes sense:
I'm blocked on problems with versions of protoc right now, I opened PRs to fix it:
https://github.com/DataDog/sketches-go/pull/19
also the decoding of mapping should be done in the sketch repo: https://github.com/DataDog/sketches-go/pull/20

while waiting on review on these PRs, I need to use this protobuf (but I will update it when my PRs are merged).

File: pkg/trace/pb/README.md (hunk 10 above)
Comment (piochelepiotr) [reply pkg/trace/pb/README.md:1]: Ok, I think the best thing would be to have some sort of makefile to generate all that (with the modifications). But it has nothing to do with that PR. I will remove it for now.

File: go.mod [hunk 16]
```diff
@@ -32,6 +32,7 @@ replace (
 
 // pinned to grpc v1.26.0
 replace (
+	github.com/golang/latestprotobuf => github.com/golang/protobuf v1.4.3
```
Comment (piochelepiotr) [inline go.mod:35]: Let me know if there is a non ugly version of doing this.

File: pkg/trace/pb/ddsketch.proto (hunk 2 above)
Comment (piochelepiotr) [reply pkg/trace/pb/ddsketch.proto]: Actually, the problem was the version of the protobuf library we were using. I changed it and I was able to remove this protobuf and use the one from the sketches repo.

File: go.mod (hunk 16 above)
Comment (gbbr) [reply go.mod:35]: Not exactly sure why you did this. I see you've pushed protobuf below to `1.4.1`. Why not straight to `1.4.3` and that's that?

File: pkg/trace/stats/quantile/converter.go [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+	"github.com/DataDog/sketches-go/ddsketch/mapping"
+	"github.com/DataDog/sketches-go/ddsketch/pb/sketchpb"
+	"github.com/golang/latestprotobuf/proto"
+)
+
+// ddSketch represents the sketch described here: http://www.vldb.org/pvldb/vol12/p2195-masson.pdf
+// This representation only supports positive values.
+type ddSketch struct {
+	// bins is the map from index --> count
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter.go:20]:
```suggestion
	// bins is the map from index to count
```
🙂

File: pkg/trace/stats/quantile/converter.go [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+	bins map[int32]float64
+	// contiguousBins is a more compact representation for contiguous bins.
+	// the index of each bin is the index in the array + contiguousBinsOffset
+	contiguousBins       []float64
+	contiguousBinsOffset int
+	// zeros is the count of 0 and its close neighbors (close defined by gamma).
+	zeros int
+	// mapping is the mapping of the ddSketch: interpolation, gamma and global index offset.
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter.go:28]: This is rocket science to me, but I trust that anyone reading the `pdf` linked above will know all of these terms so thanks for documenting 👍

File: pkg/trace/stats/quantile/converter.go [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+	return count
+}
+
+func (s *ddSketch) maxSize() int {
+	return len(s.bins) + len(s.contiguousBins)
+}
+
+// getIndexes returns all the sorted indexes contained in s1 and s2.
+func getIndexes(s1 ddSketch, s2 ddSketch) []int {
+	// todo: No need to re-allocate that array at each conversion.
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter.go:49]:
```suggestion
	// TODO(x): No need to re-allocate that array at each conversion.
```
Nit: Usually we use `TODO(github_username)` or `TODO(x)` if anyone else can pick it up.

File: pkg/trace/stats/quantile/converter.go [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+	// but this function needs to be thread safe in the agent.
+	indexes := make([]int, 0, s1.maxSize()+s2.maxSize())
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter.go:51]: Just a question. What if we used `is` (basically the plural of `i`), and then below instead of `ind` we would use `newi`? It'd be shorted and perhaps easier to follow but I don't feel strongly and I'm fine if you want to leave it exactly as it is.

File: pkg/trace/stats/quantile/converter.go [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+	for i := range s1.contiguousBins {
+		indexes = append(indexes, i+s1.contiguousBinsOffset)
+	}
+	for i := range s2.contiguousBins {
+		ind := i + s2.contiguousBinsOffset
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter.go:56]: IMHO `ind` should be `index` or `indexes` should become `inds` 🙂

File: pkg/trace/stats/quantile/converter_test.go [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+		} else {
+			rank := math.Ceil(testQuantiles[i] * float64(2*n))
+			exp = rank - 1
+		}
+		assert.InDelta(exp, v, EPSILON*float64(2*n)+relativeValueError*exp, "quantile %f failed, exp: %f, val: %f", testQuantiles[i], exp, v)
+	}
+}
+
+func TestDDToGKSketch(t *testing.T) {
+	t.Run("uniform10", func(t *testing.T) {
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter_test.go:121]: You could make `testDDSketchToGKUniform` to be:
```go
func testDDSketchToGKUniform(n int) func(t *testing.T) {
    return func(t *testing.T) {
        // t is available now
    }
}
```
And then here simply do:
```go
t.Run("uniform10", testDDSketchToGKUniform)
```
It's a bit nicer.

File: pkg/trace/stats/quantile/converter_test.go [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// getConvertedSketchQuantiles follows this steps:
+// 1. Generates two DDSketches: OK and Errors:
+//    - Errors uses the generator function from 0 to n-1
+//    - OK uses the generator function from n to 2*n-1)
+// 2. Converts OK and errors DDSketches to hits and errors GK sketches (hits = ok + errors)
+//    - That way, hits is the distribution from 0 to 2*n-1
+// 3. Computes quantiles on the GK sketches hits and errors and returns them
+func getConvertedSketchQuantiles(t *testing.T, n int, gen func(i int) float64, testQuantiles []float64) (hits []float64, errors []float64) {
```
Comment (gbbr) [inline pkg/trace/stats/quantile/converter_test.go:41]: I don't want to force your hand but normally, if possible, I'd expect a suite of tests for this function, since all tests rely on it. I haven't analysed it much so if you think it's not needed it's fine to leave it out. However, if at any time the tests break, it will be unfortunate to discover that it's because of a helper method...

**Review by gbbr — COMMENTED** [review]
Looking great. Thanks for taking the time! This is definitely not a draft anymore.

File: go.mod (hunk 16 above)
Comment (piochelepiotr) [reply go.mod:35]: we still have `github.com/golang/protobuf => github.com/golang/protobuf v1.3.2` so I don't know if this is done on purpose to stay on the old version. If I can bump that, that would be perfect.

File: pkg/trace/stats/quantile/converter_test.go (hunk 22 above)
Comment (piochelepiotr) [reply pkg/trace/stats/quantile/converter_test.go]: Nice!

File: go.mod (hunk 16 above)
Comment (gbbr) [reply go.mod:35]: I'd suggest to `git blame` that line and ask the person. It should likely be fine.

File: pkg/trace/stats/quantile/ddsketch_converter.go (hunk 14 above)
Comment (gbbr) [reply pkg/trace/stats/quantile/ddsketch_converter.go]: All good then.

File: pkg/trace/stats/quantile/converter_test.go (hunk 23 above)
Comment (piochelepiotr) [reply pkg/trace/stats/quantile/converter_test.go:41]: The function is a wrapper around the function that we test. So not sure what we can test that we are not already testing?
I thought about testing quantiles on the DDSketch before converting it, but it's not exported yet in the `sketches-go` library (I opened a PR to export it: https://github.com/DataDog/sketches-go/pull/22)

File: pkg/trace/stats/quantile/converter.go [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	if hits.N > 0 {
+		hits.Entries[0].Delta = 0
+		hits.Entries[len(hits.Entries)-1].Delta = 0
+	}
+	if errors.N > 0 {
+		errors.Entries[0].Delta = 0
+		errors.Entries[len(errors.Entries)-1].Delta = 0
+	}
+	hits.compress()
```
Comment (furmmon) [inline pkg/trace/stats/quantile/converter.go:168]: I think we can remove compression step. ddsketch already controls and reduce the size of the sketches.
Not compressing would a allow a conversion back to ddsketch without any loss. Allowing services to consider these as ddsketches or manually compress

**Review by raphaelgavache — APPROVED** [review]
lgtm for conversion

File: go.mod [hunk 25]
```diff
@@ -103,7 +104,7 @@ require (
 	github.com/gogo/protobuf v1.3.1
 	github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
 	github.com/golang/mock v1.3.1
-	github.com/golang/protobuf v1.3.4
+	github.com/golang/protobuf v1.4.1
```
Comment (gbbr) [inline go.mod:107]:
```suggestion
	github.com/golang/protobuf v1.4.3
```
Why not like this? And remove the `replace` clause

**Review by gbbr — COMMENTED** [review]
All good to me, just one small question.

**Review by gbbr — APPROVED** [review]

File: pkg/trace/test/testsuite/testdata/clientstats.go [hunk 26]
```diff
@@ -3,8 +3,25 @@ package testdata
...
 	"github.com/DataDog/datadog-agent/pkg/trace/stats"
+	"github.com/DataDog/datadog-agent/pkg/trace/stats/quantile"
+
+	"github.com/DataDog/sketches-go/ddsketch"
+	"github.com/DataDog/sketches-go/ddsketch/mapping"
+	"github.com/DataDog/sketches-go/ddsketch/store"
+	"github.com/golang/protobuf/proto"
 )
 
+func getEmptyDDSketch() []byte {
```
Comment (piochelepiotr) [inline pkg/trace/test/testsuite/testdata/clientstats.go:14]: In this file, we are only testing if distributions are created with the correct keys. That's why we use empty sketches. We test the actual conversion in `conversion_test.go`
